src/VisualStudio/Core/Def/Implementation/Interop/CleanableWeakComHandleTable.cs

```diff
@@ -39,135 +42,88 @@ internal class CleanableWeakComHandleTable<TKey, TValue> : ForegroundThreadAffin
-                var key = _deadKeySet.First();
-
-                _deadKeySet.Remove(key);
-
-                Debug.Assert(_table.ContainsKey(key), "Key not found in table.");
-                _table.Remove(key);
-
-                if (timeSlice.IsOver)
-                {
-                    return false;
-                }
-            }
-
-            return true;
         }
 
         /// <summary>
-        /// Cleans up references to dead objects in the table. This operation will return if it takes
-        /// longer than <see cref="CleanUpTimeSlice"/>. Calling <see cref="CleanUpDeadObjects"/> further
-        /// times will continue the process.
+        /// Cleans up references to dead objects in the table. This operation will yield to other foreground operations
+        /// any time execution exceeds <see cref="CleanUpTimeSlice"/>.
         /// </summary>
-        public void CleanUpDeadObjects()
+        public async Task CleanUpDeadObjectsAsync()
         {
-            this.AssertIsForeground();
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(ThreadingContext.DisposalToken);
 
             if (!_needsCleanUp)
             {
                 return;
             }
 
+            // Immediately mark as not needing cleanup; this operation will clean up the table by the time it returns.
+            _needsCleanUp = false;
+
             var timeSlice = new TimeSlice(CleanUpTimeSlice);
 
-            if (_cleanUpState == CleanUpState.Initial)
-            {
-                _cleanUpEnumerator = _table.GetEnumerator();
-                _cleanUpState = CleanUpState.CollectingDeadKeys;
-            }
+            await CollectDeadKeysAsync().ConfigureAwait(true);
+            await RemoveDeadKeysAsync().ConfigureAwait(true);
+            return;
 
-            if (_cleanUpState == CleanUpState.CollectingDeadKeys)
+            // Local functions
+            async Task CollectDeadKeysAsync()
             {
-                if (_cleanUpEnumerator == null)
+                var cleanUpEnumerator = _table.GetEnumerator();
+                while (cleanUpEnumerator.MoveNext())
                 {
-                    // The enumerator got reset while we were collecting dead keys.
-                    // Go ahead and remove the dead keys that were already collected before
-                    // collecting more.
-                    if (!RemoveDeadKeys(timeSlice))
+                    var pair = cleanUpEnumerator.Current;
+                    if (!pair.Value.IsAlive())
                     {
-                        return;
-                    }
+                        _deadKeySet.Add(pair.Key);
 
-                    _cleanUpEnumerator = _table.GetEnumerator();
-                }
+                        if (timeSlice.IsOver)
+                        {
+                            await ResetTimeSliceAsync().ConfigureAwait(true);
+                            await RemoveDeadKeysAsync().ConfigureAwait(true);
```

Comment:
these are the sorts of parts i'd just like comments on how this will be safe even if another CleanUpDeadObjectsAsync call ran.

Reply:
Comments added.